src/Grpc.Net.Client/Internal/Retry/HedgingCall.cs

```diff
@@ -60,151 +60,157 @@ public HedgingCall(HedgingPolicyInfo hedgingPolicy, GrpcChannel channel, Method<
                     }
+                    else
+                    {
+                        CommitCall(call, CommitReason.FatalStatusCode);
+                    }
                 }
-                else
-                {
-                    CommitCall(call, CommitReason.FatalStatusCode);
-                }
-            }
 
-            lock (Lock)
-            {
-                if (IsDeadlineExceeded())
-                {
-                    // Deadline has been exceeded so immediately commit call.
-                    CommitCall(call, CommitReason.DeadlineExceeded);
-                }
-                else if (_activeCalls.Count == 1 && AttemptCount >= MaxRetryAttempts)
-                {
-                    // This is the last active call and no more will be made.
-                    CommitCall(call, CommitReason.ExceededAttemptCount);
-                }
-                else if (_activeCalls.Count == 1 && IsRetryThrottlingActive())
-                {
-                    // This is the last active call and throttling is active.
-                    CommitCall(call, CommitReason.Throttled);
-                }
-                else
+                lock (Lock)
                 {
-                    // Call isn't used and can be cancelled.
-                    // Note that the call could have already been removed and disposed if the
-                    // hedging call has been finalized or disposed.
-                    if (_activeCalls.Remove(call))
+                    if (IsDeadlineExceeded())
                     {
-                        call.Dispose();
+                        // Deadline has been exceeded so immediately commit call.
+                        CommitCall(call, CommitReason.DeadlineExceeded);
+                    }
+                    else if (_activeCalls.Count == 1 && AttemptCount >= MaxRetryAttempts)
+                    {
+                        // This is the last active call and no more will be made.
+                        CommitCall(call, CommitReason.ExceededAttemptCount);
+                    }
+                    else if (_activeCalls.Count == 1 && IsRetryThrottlingActive())
+                    {
+                        // This is the last active call and throttling is active.
+                        CommitCall(call, CommitReason.Throttled);
+                    }
+                    else
+                    {
+                        // Call isn't used and can be cancelled.
+                        // Note that the call could have already been removed and disposed if the
+                        // hedging call has been finalized or disposed.
+                        if (_activeCalls.Remove(call))
+                        {
+                            call.Dispose();
+                        }
                     }
                 }
             }
-
-            if (CommitedCallTask.IsCompletedSuccessfully() && CommitedCallTask.Result == call)
+            finally
```

Comment:
I noticed we don't do any except handling on the top-level of the `StartCall` here. Is it worthwhile to do here or is the exception handling at the call sites sufficient?

Reply:
All logic that can throw should be inside a try/catch.

However, a global catch couldn't hurt. Also, throwing on CTS cancellation and reusing `HandleUnexpectedError` simplifies the code.

Updated.